src/main/java/com/google/devtools/build/lib/remote/http/HttpDownloadHandler.java

```diff
@@ -105,8 +116,18 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws Ex
 
       ByteBuf content = ((HttpContent) msg).content();
       int readableBytes = content.readableBytes();
-      content.readBytes(out, readableBytes);
+      int readerIndex = content.readerIndex();
+      try {
+        content.readBytes(out, readableBytes);
+      } catch (ZipException ex) {
+        // not a compressed response, fall back to uncompressed
+        logger.log(Level.WARNING, "Decompression failed: " + ex.toString());
+        content.readerIndex(readerIndex);
+        out = rawOut;
+        content.readBytes(out, readableBytes);
+      }
```

Comment:
Can we handle this in `HttpCacheClient` based on the option value rather than passing it down to the handler?

Reply:
We can wrap the output stream there, yes. May be a bit fiddly to get the fallback working though. Will take a look.